### DISCUSSION TITLE
Native image distribution skipped because there are "no matching artifacts"?

### DISCUSSION THREAD

**Discussion by klausbrunner** [jreleaser/jreleaser#1220] [Q&A]
Trying out JReleaser 1.5.0 for #1196, and I'm wondering if I've misunderstood the change. I'm building native images for 3 VM types on a GH Actions matrix build, which works fine. I then proceed to collect the target artifacts from all those machines in a separate job and run jreleaser:full-release.

So far, I've had to explicitly list the artifacts to pick up in the distribution, which worked but was a bit verbose. My understanding was that thanks to #1196, these would be picked up automatically, but I'm getting:
```
Warning:  == JReleaser ==
Warning:  distribution.solarpos-native was disabled because there are no matching artifacts
```
Here's the relevant config for the assembly:

```xml
                            <nativeImage>
                                <solarpos-native>
                                    <active>ALWAYS</active>
                                    <executable>${project.artifactId}</executable>
                                    <mainJar>
                                        <path>target/${project.artifactId}-${project.version}.jar</path>
                                    </mainJar>
                                    <args>
                                        <arg>--no-fallback</arg>
                                        <arg>--verbose</arg>
                                    </args>
                                </solarpos-native>
                            </nativeImage>
```

And here's how target/jreleaser looks on the machine that's running jreleaser:full-release (just before starting it):

```
    ├── jreleaser
    │   ├── assemble
    │   │   ├── solarpos
    │   │   │   └── java-archive
    │   │   │       ├── solarpos-0.1.2-SNAPSHOT.zip
    │   │   │       └── work
    │   │   │           └── solarpos-0.1.2-SNAPSHOT
    │   │   │               ├── LICENSE
    │   │   │               ├── bin
    │   │   │               │   ├── solarpos
    │   │   │               │   └── solarpos.bat
    │   │   │               └── lib
    │   │   │                   └── solarpos-0.1.2-SNAPSHOT.jar
    │   │   └── solarpos-native
    │   │       └── native-image
    │   │           ├── jars
    │   │           │   └── universal
    │   │           │       └── solarpos-0.1.2-SNAPSHOT.jar
    │   │           ├── solarpos
    │   │           ├── solarpos-0.1.2-SNAPSHOT-linux-x86_64.zip
    │   │           ├── solarpos-0.1.2-SNAPSHOT-osx-x86_64.zip
    │   │           ├── solarpos-0.1.2-SNAPSHOT-windows-x86_64.zip
    │   │           ├── solarpos.build_artifacts.txt
    │   │           └── solarpos.exe
    │   ├── output.properties
    │   └── trace.log
```

I would have expected that the ...linux/osx/windows-x86_64-zip files would be found automatically and made part of the distribution, but that's not happening. Am I missing some crucial config here?

**Comment by aalmiray** [comment]
I think the missing aspect is defining `<graalJdks>` entries, otherwise the assembler won't know there are multiple archives. It would only export a single archive matching the current JVM (assumed to be a GraalVM JDK).

**Reply by aalmiray** [reply]
Indeed. Explicitly defining target GraalJdks will automatically populate potential archives as seen at https://github.com/jreleaser/jreleaser/blob/e2d815d708f37f1d86c21a17125e13b72ff789d1/core/jreleaser-model-impl/src/main/java/org/jreleaser/model/internal/validation/assemble/NativeImageAssemblerResolver.java#L62-L85

However, there's a bug. Line 63 will automatically filter all artifacts not matching the current platform. Which means no matter how many GraalJdks you may define you'll only get a single candidate archive. I thought I had test this settings, but this bug managed to get through the cracks.

**Reply by aalmiray** [reply]
filed https://github.com/jreleaser/jreleaser/issues/1221

**Reply by aalmiray** [reply]
Ugh. Actually #1221 is invalid. It does work. The real problem here are the missing `<graalJdks>`. I'm actually working on a solution right now for https://github.com/jreleaser/helloworld-java-graalvm

**Comment by klausbrunner** [comment]
OK, I see. But in this case, defining graalJdks would be somewhat fake, because they don't actually exist on the machine doing the distribution - all I'd be saying is "you should expect artifacts for these platforms". Which is fine too, I just feel there is a bit of a logical disconnect between the assemble step being able to build the relevant artifacts on its own just fine, but the distribution step then not knowing anything about them.

Naive half-baked idea: have a "bill of materials" directory in target that contains one uniquely-named text file for each artifact that's built by the assembly step(s), containing its actual name. That way, the BOM could be safely accumulated by the build nodes without overwriting each other's output and the distribution step knows exactly which files to use.

Or if that's too complex or too implicit, introduce an alternative config to graalJDKs that basically justs lists the expected platforms for the distribution step. If they aren't all there, bail out. That way, the config isn't cluttered with redudant file name expressions and it's not bound to things that only the assemble step should need to know.

**Reply by aalmiray** [reply]
Listing just the expected platforms might work. Have to think about the implications of it though.

**Reply by aalmiray** [reply]
Listing platforms may not be enough as one can override the archive extension by setting extra properties on a target GraalJdk. See https://github.com/jreleaser/jreleaser/blob/e2d815d708f37f1d86c21a17125e13b72ff789d1/core/jreleaser-model-impl/src/main/java/org/jreleaser/model/internal/validation/assemble/NativeImageAssemblerResolver.java#L67-L70

**Reply by aalmiray** [reply]
Alright. I get where you're coming from.

One the one hand listing multiple `<graalJdks>` in the assembler looks "fake" because only a single GraalJdk can be active for a given platform. Thus when running on Windows the assembler doesn't care about Linux or OSX. Not listing a single `<graalJdk>` forces the assembler to use the current JDK used to run the build and hope it's a matching GraalVM JDK. Listing candidate GraalVM JDKs lets you decouple the build/release JDK from the assembler JDK, so that if the release is run with say Java 17 you could still assemble a native-image with a matching GraalJDK 17. Listing `<graalJdks>` gives the assembler the option to fill the list of candidate artifacts if the assembly is exported as a distribution.

Not listing `<graalJdks>` in the assembler forces it to just build a single archive matching the current platform. for the same reason it can only resolve an export a single artifact. The alternative is to explicitly list all candidate artifacts with a matching `distribution`. This is akin to the BOM like idea you suggested if I understood it correctly, as the distribution lets you define an arbitrary list of artifacts.

**Reply by klausbrunner** [reply]
Specifying a separate GraalVM JDK makes complete sense, it's just not what I would do on a Github Matrix build (or similar systems like Travis CI). I'll just setup GraalVM as the only JDK on the particular build machine and let it build the JVM and native parts, without fiddling around with multiple JDK installations. This works fine and makes both the workflow job and JReleaser config fairly straightforward. For illustration, my workflow yaml is [this](https://github.com/klausbrunner/solarpos/blob/main/.github/workflows/build-and-release.yml).

So, the BOM idea is to decide the artifacts for distribution at build time, rather than having to pre-define all the distribution platforms in the JReleaser config: the assemblies would "push" the information about the actually built artifacts to the distribution. My motivation here is to define the list of platforms just once (in my case, in the Github workflow yaml) and jreleaser would then pick up whatever was actually built by the matrix builds.

Obviously, all of this would be a lot easier if GraalVM could do cross-compilation, alas...

**Reply by aalmiray** [reply]
I think this use case is close to:

- configuring the native-image assembler with `exported: false`. This way no matching distribution is configured with the single archive that the resolver should find.
- explicitly configuring a distribution with all candidate artifacts.
- CI workflow must place candidate artifacts at the expected location. Note that `environment:` properties may be used to override said location, as it's done with [kcctl](https://github.com/kcctl/kcctl/blob/3ee3166f15252a6a01ac095d5d9db0b095b9643c/pom.xml#L484-L518) for example
